ucsmsdk/utils/comparesyncmo.py

```diff
@@ -0,0 +1,489 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains the api used to provide compare and sync functionality
+within ucsm or across ucsm.
+"""
+
+from __future__ import print_function
+from __future__ import absolute_import
+
+import re
+import logging
+
+from .. import ucsgenutils
+from .. import ucscoreutils
+from ..ucscoremeta import MoPropertyMeta, MoMeta
+from ..ucsexception import UcsWarning, UcsValidationException
+from ..ucsmo import GenericMo
+
+log = logging.getLogger('ucs')
+
+skip_mos = ["LsVersionBeh", "LsmaintAck", "LsIssues", "VnicIpV4PooledAddr",
+            "VnicDynamicCon", "LsVConAssign", "ComputePooledSlot",
+            "ComputePooledRackUnit", "VnicIPv4PooledIscsiAddr",
+            "IppoolPooled", "IqnpoolPooled", "MacpoolPooled",
+            "UuidpoolPooled", "VnicConnDef", "VnicFcGroupDef",
+            "LsServerExtension"]
+
+
+class _CompareStatus(object):
+    """
+    Internal class used as enum.
+    """
+
+    TYPES_DIFFERENT = 0
+    PROPS_DIFFERENT = 1
+    EQUAL = 2
+
+
+class _PropDiff(object):
+    """
+    Internal class for property delta
+    """
+    def __init__(self, prop, old_value, new_value):
+        self.prop = prop
+        self.old_value = old_value
+        self.new_value = new_value
+
+
+class _MoDiff(object):
+    """
+    This class represents difference object.
+    """
+
+    REMOVE = "<="
+    ADD_MODIFY = "=>"
+    EQUAL = "=="
+
+    def __init__(self, input_object, side_indicator, diff_property=None,
+                 ref_values=None, diff_values=None):
+        self.input_object = input_object
+        self.dn = input_object.dn
+        self.side_indicator = side_indicator
+        self.diff_property = diff_property
+
+        if ref_values:
+            self.ref_prop_values = ref_values
+        else:
+            self.ref_prop_values = {}
+
+        if diff_values:
+            self.diff_prop_values = diff_values
+        else:
+            self.diff_prop_values = {}
+
+    def get_prop_diff(self, prop):
+        if prop in self.ref_prop_values and prop in self.diff_prop_values:
+            prop_name = ""
+            for key in self.ref_prop_values:
+                if prop.lower() == key.lower():
+                    prop_name = key
+                    break
+            return _PropDiff(prop_name, self.ref_prop_values[prop],
+                             self.diff_prop_values[prop])
+        return None
+
+
+def _compare(from_mo, to_mo, diff, include_operational=False):
+    """
+    Internal method to support compare reference and difference object.
+    """
+
+    if from_mo.get_class_id() != to_mo.get_class_id():
+        return _CompareStatus.TYPES_DIFFERENT
+
+    # compare for unknown class_id
+    if isinstance(from_mo, GenericMo) or isinstance(to_mo, GenericMo):
+        for prop in from_mo.properties:
```

Comment:
@ragupta-git I assume that `.properties` will be available in a regular Mo also? And this will not fail if one mo is a `gmo` and other is a regular mo. We should add a test case for this